### PR TITLE
Enable purely iteration-based training

### LINKED ISSUE #3521: Enable training purely based on number of iterations instead of epochs
## 🚀 Feature
Enable training purely based on number of iterations instead of epochs

### Motivation

This can be useful for certain training runs. Without this feature, the user must set an unreachably high value for `max_epochs` and set `max_steps` to the desired iteration count. With this setup, the trainer will break from the training loop based on `max_steps` since we'd never reach `max_epochs`. For example, [Detectron2](https://github.com/facebookresearch/detectron2/blob/master/detectron2/engine/train_loop.py#L125-L147) uses iteration-based training in their train loop.

### Pitch
The solution could be pretty simple. We can make `min_epochs` and `max_epochs` as Optional. If all of min/max_epochs and max_steps are unset, use the defaults we have today (min_epochs=1, max_epochs=1000). If only `max_steps` is set, use that (and keep `min/max_epochs` as None). If all are set, stop based on whichever condition is hit first.

Specifically, we'd need to:
- Initialize things correctly here: https://github.com/PyTorchLightning/pytorch-lightning/blob/c64520e658806a87f282c74299b2ea4b7ea493ea/pytorch_lightning/trainer/training_loop.py#L46
- Update the training loop here to not depend on `max_epochs`: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L340-L375 . 

Are there other spots I'm missing? 

### Alternatives

Without touching the Trainer code, a super hacky solution would be setting `max_epochs` to be an outrageously large value and set `max_steps` to my desired iteration count. then we'll break from the train loop according to `max_steps` since we'd meet this condition first instead of `max_epochs`. However this could be better handled

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #3521

This preserves the existing default of `min_epochs=1` if both `min_steps` and `min_epochs` aren't specified, and the default behavior of `max_epochs=1000` if both `max_steps` and `max_epochs` aren't specified.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
this is the same as #5686 #5687 correct?

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -71,8 +71,10 @@ def on_trainer_init(
         self.trainer.train_dataloader = None
         self.automatic_optimization = automatic_optimization
 
-        self.trainer.max_epochs = max_epochs
-        self.trainer.min_epochs = min_epochs
+        # If neither max_epochs or max_steps is set, then use existing default of max_epochs = 1000
+        self.trainer.max_epochs = 1000 if (max_epochs is None and max_steps is None) else max_epochs
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:75]: could be a constant

File: tests/trainer/flags/test_min_max_epochs.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+@pytest.mark.parametrize(
+    ["min_epochs", "max_epochs", "min_steps", "max_steps"],
+    [
+        pytest.param(None, 5, None, None),
+        pytest.param(None, None, None, 100),
+        pytest.param(None, 5, None, 100),
+        pytest.param(None, None, 10, 100),
+        pytest.param(1, 5, None, None),
+        pytest.param(1, None, None, 100),
+        pytest.param(None, 5, 10, None),
```
Comment (awaelchli) [inline tests/trainer/flags/test_min_max_epochs.py:20]: could we lower the numbers just a little bit to speed up the test?

File: tests/trainer/flags/test_min_max_epochs.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import pytest
+
+from pytorch_lightning import Trainer
+from tests.base.boring_model import BoringModel
+
+
+# @pytest.mark.parametrize("min_epochs", [None, 2])
+# @pytest.mark.parametrize("max_epochs", [None, 3])
+# @pytest.mark.parametrize("min_steps", [None, 20])
+# @pytest.mark.parametrize("max_steps", [None, 100])
```
Comment (awaelchli) [inline tests/trainer/flags/test_min_max_epochs.py:10]: Are all of these 16 configurations valid?
Did you try what happens if we choose max_epochs = 1 and max_steps = 5 but the epoch only has 3 batches?
The number that leads to less training steps should terminate training, right? This could be a separate test.

**Review by awaelchli — COMMENTED** [review]
nice!

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -101,8 +102,8 @@ def __init__(
         check_val_every_n_epoch: int = 1,
         fast_dev_run: Union[int, bool] = False,
         accumulate_grad_batches: Union[int, Dict[int, int], List[list]] = 1,
-        max_epochs: int = 1000,
-        min_epochs: int = 1,
+        max_epochs: Optional[int] = None,
+        min_epochs: Optional[int] = None,
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:106]: this is confusing, you say there is a default number but you make it as None
in general None default is used for deprecated or mutable values (such as list)

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:106]: that's fair. if I set a default for min/max_epochs, then users who want to do training based purely on steps need to specify this:
```
trainer = Trainer(min_steps=M, max_steps=N, min_epochs=None, max_epochs=None, ...)
```

instead they can simply specify this:
```
trainer = Trainer(min_steps=M, max_steps=N, ...)
```

so by setting these defaults to `None` in the constructor and handling the initialization elsewhere, we save users some of the boilerplate code from having to null out the epoch arguments if they don't use them. what do you think?

File: tests/trainer/flags/test_min_max_epochs.py (hunk 3 above)
Comment (ananthsub) [reply tests/trainer/flags/test_min_max_epochs.py:10]: if both are specified, i think should exit out whenever the first condition is met. so for the example you provided, if the epoch has only 3 batches, i agree we should return after the epoch completes and before moving onto the remaining 2 steps. the same goes for min_steps

i think users can already exercise this functionality now, so i'll look to see if there are already test cases for it. if not i'll add them here

**Comment by ananthsub** [comment]
> this is the same as #5686 #5687 correct?

yes that's correct, i'm sorry about the noisy PRs. i made some n00b git errors and it was easier for me to create a new PR than fix the existing ones

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:75]: Could be but no real need. You can set it with max_epochs

Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:75]: I mean, it's not like in the ModelCheckpoint constants, where you'd need to hack it to change the last checkpoint name if SAVE_LAST_NAME didn't exist

File: tests/trainer/flags/test_min_max_epochs.py (hunk 2 above)
Comment (carmocca) [reply tests/trainer/flags/test_min_max_epochs.py:20]: Use plain tuples instead of `pytest.param` if its advantages aren't necessary

**Comment by ananthsub** [comment]
one thing i need to check is what happens to the progress bar

File: tests/trainer/flags/test_min_max_epochs.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    result = trainer.fit(model)
+    assert result == 1, "Training did not complete"
+
+    # check training stopped at max_epochs or max_steps
+    if trainer.max_steps and not trainer.max_epochs:
+        assert trainer.global_step == trainer.max_steps
+
+    if trainer.max_steps and not trainer.max_epochs:
+        assert trainer.global_step == trainer.max_steps
```
Comment (kaushikb11) [inline tests/trainer/flags/test_min_max_epochs.py:46]: Code repetition?

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -599,7 +603,7 @@ def train(self):
                     return
 
                 # early stopping
-                met_min_epochs = epoch >= self.min_epochs - 1
+                met_min_epochs = epoch >= self.min_epochs - 1 if self.min_epochs else True
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:606]:
```suggestion
                met_min_epochs = (epoch >= self.min_epochs - 1) if self.min_epochs else True
```

File: pytorch_lightning/trainer/training_loop.py [hunk 7]
```diff
@@ -93,7 +95,9 @@ def num_optimizers(self):
         return num_optimizers
 
     def should_skip_training(self):
-        return self.trainer.current_epoch >= self.trainer.max_epochs or self.trainer.num_training_batches == 0
+        return (
+            self.trainer.max_epochs is not None and self.trainer.current_epoch >= self.trainer.max_epochs
+        ) or self.trainer.num_training_batches == 0
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py:100]:
```suggestion
        should_by_epoch = self.trainer.max_epochs is not None and self.trainer.current_epoch >= self.trainer.max_epochs
        return should_by_epoch or self.trainer.num_training_batches == 0
```

File: tests/trainer/flags/test_min_max_epochs.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import pytest
+
+from pytorch_lightning import Trainer
+from tests.helpers import BoringModel
+
+
+# @pytest.mark.parametrize("min_epochs", [None, 2])
+# @pytest.mark.parametrize("max_epochs", [None, 3])
+# @pytest.mark.parametrize("min_steps", [None, 20])
+# @pytest.mark.parametrize("max_steps", [None, 100])
```
Comment (Borda) [inline tests/trainer/flags/test_min_max_epochs.py:10]:
```suggestion
```

**Review by kaushikb11 — APPROVED** [review]

File: tests/trainer/flags/test_min_max_epochs.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+@pytest.mark.parametrize(
+    ["min_epochs", "max_epochs", "min_steps", "max_steps"],
+    [
+        pytest.param(None, 5, None, None),
+        pytest.param(None, None, None, 100),
+        pytest.param(None, 5, None, 100),
+        pytest.param(None, None, 10, 100),
+        pytest.param(1, 5, None, None),
+        pytest.param(1, None, None, 100),
+        pytest.param(None, 5, 10, None),
```
Comment (carmocca) [inline tests/trainer/flags/test_min_max_epochs.py:16]:
```suggestion
        (None, 3, None, None),
        (None, None, None, 20),
        (None, 3, None, 20),
        (None, None, 10, 20),
        (1, 3, None, None),
        (1, None, None, 20),
        (None, 3, 10, None),
```

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
